diff-containers/README.md

```diff
@@ -1 +1,238 @@
+there is only one such implemented difference datatype for `Data.Map.Strict`.
+
+## Differences on `Data.Map`s
+
+A difference datatype for maps is defined and implemented in the
+`Data.Map.Diff.Strict` module. A key property of our definition of diffs is that
+diffs can not only be combined/summed, they can also be inverted and summed
+together, which facilitates *subtraction* of diffs. Our definition of diffs is
+not only a `Semigroup` and a `Monoid`, it is also a `Group`.
+
+The goals for our definition of diffs are twofold:
+1. Provide definitions that adhere to the type class laws for `Semigroup`,
+   `Monoid` and `Group`.
+2. Applying diffs to maps should have a nice semantics.
+
+An overview of the sections in this document:
+
+* Section [1) High-level overview](#high-level-overview) provides a high-level
+  overview of diffs, how to create them, and how to apply them.
+* Section [2) A first attempt at implementing
+  diffs](#2-a-first-attempt-at-implementing-diffs) describes a straightforward
+  way to implement diffs, and where this goes wrong if we try to define a
+  `Group` instance.
+* Section [3) Implementing diffs with
+  inverses](#3-implementing-diffs-with-inverses) describes how we implement
+  diffs to facilitate a `Group` instance.
+* Section [4) Applying diffs with inverses: positivity and
+  normality](#4-applying-diffs-with-inverses-positivity-and-normality) describes
+  how applying diffs can fail, but which can be solved by guaranteeing the
+  positivity and normality properties.
+
+### 1) High-level overview
+
+For each of the key-value pairs in a map, the difference type will keep track of
+whether a value is to be inserted or deleted:
+
+```haskell
+newtype Diff k v = Diff {- omitted -}
+
+data DiffEntry v =
+    Insert !v
+  | Delete !v
+  {- omitted for now -}
+```
+
+One method of constructing a diff is to compute the difference between two maps.
+Other methods include creating diffs from maps and lists directly:
+
+```haskell
+diff :: (Ord k, Eq v) => Map k v -> Map k v -> Diff k v
+
+fromMap :: Map k (DiffEntry v) -> Diff k v
+fromList :: Ord k => [(k, DiffEntry v)] -> Diff k v
+```
+
+Diffs can be applied to maps to obtain new maps. For diffs constructed using
+`diff`, the following `property` also holds:
+
+```haskell
+unsafeApplyDiff :: Ord k => Map k v -> Diff k v -> Map k v
+
+property :: Map k v -> Map k v -> Bool
+property = unsafeApplyDiff m1 (diff m1 m2) == m2
+```
+
+### 2) A first attempt at implementing diffs
+
+A straightfoward to represent differences on maps is to keep track of the most
+recent change for each key in the map.
+
+```haskell
+type Diff_ k v = Diff_ (Map k (Delta_ v))
+data Delta_ = Insert_ !v | Delete_ !v
+```
+
+Diffs can then be summed (or combined/`mappend`ed) through a
+`Semigroup`/`Monoid` instance that behaves like a pair-wise `Last` monoid.
+
+```haskell
+   fromList [(1, Insert_ 54), (2, DDelete 42)                ]
```

Comment:
I think it's the first time we use `DDelete`. Shouldn't we introduce it first?

Reply:
My mistake, it should be `Delete_`!